src/utils/layout-helpers.ts

```diff
@@ -139,46 +140,90 @@ export function isELSAddonRoot(root: string) {
   return hasEmberLanguageServerExtension(pack);
 }
 
-export function getProjectInRepoAddonsRoots(root: string) {
-  const prefix = isModuleUnificationApp(root) ? 'packages' : 'lib';
-  const addons = safeWalkSync(path.join(root, prefix), {
-    directories: true,
-    globs: ['**/package.json'],
-  });
-  const engineAddons = safeWalkSync(path.join(root, 'engines'), {
-    directories: true,
-    globs: ['**/package.json'],
-  });
-  const roots: string[] = [];
+function getRecursiveInRepoAddonRoots(root: string, roots: string[]) {
+  const packageData = getPackageJSON(root);
+  const emberAddonPaths: string[] = (packageData['ember-addon'] && packageData['ember-addon'].paths) || [];
+
+  if (roots.length) {
+    if (!isEmberAddon(packageData)) {
+      return [];
+    }
+  }
 
-  addons
-    .map((relativePath: string) => {
-      return path.dirname(path.join(root, prefix, relativePath));
+  const recursiveRoots: string[] = roots.slice(0);
+
+  emberAddonPaths
+    .map((relativePath) => path.normalize(path.join(root, relativePath)))
+    .filter((packageRoot: string) => {
+      return isProjectAddonRoot(packageRoot);
     })
-    .filter((packageRoot: string) => isProjectAddonRoot(packageRoot))
     .forEach((validRoot: string) => {
-      roots.push(validRoot);
-      getProjectAddonsRoots(validRoot, roots).forEach((relatedRoot: string) => {
-        if (!roots.includes(relatedRoot)) {
-          roots.push(relatedRoot);
-        }
-      });
+      const packInfo = getPackageJSON(validRoot);
+
+      // we don't need to go deeper if package itself not an ember-addon or els-extension
+      if (!isEmberAddon(packInfo) && !hasEmberLanguageServerExtension(packInfo)) {
+        return;
+      }
+
+      if (!recursiveRoots.includes(validRoot)) {
+        recursiveRoots.push(validRoot);
+        getRecursiveInRepoAddonRoots(validRoot, recursiveRoots).forEach((relatedRoot: string) => {
+          if (!recursiveRoots.includes(relatedRoot)) {
+            recursiveRoots.push(relatedRoot);
+          }
+        });
+      }
     });
 
-  engineAddons
-    .map((relativePath: string) => {
-      return path.dirname(path.join(root, 'engines', relativePath));
-    })
-    .filter((packageRoot: string) => isProjectAddonRoot(packageRoot))
-    .forEach((validRoot: string) => {
-      roots.push(validRoot);
-      getProjectAddonsRoots(validRoot, roots).forEach((relatedRoot: string) => {
-        if (!roots.includes(relatedRoot)) {
-          roots.push(relatedRoot);
-        }
+  return recursiveRoots;
+}
+
+export function getProjectInRepoAddonsRoots(root: string) {
+  const roots: string[] = [];
+
+  if (!isModuleUnificationApp(root)) {
+    ['lib', 'engines'].forEach((prefix) => {
```

Comment:
looke like we don't need it for nor module unification app, it should be like:

```js
getProjectInRepoAddonsRoots(root) {
   if (isModuleUnificationApp(root)) {
     // ...legacy logic
   } else {
     // new logic, based on package.json 
   }
}

```

currently:

```js
if (not(moduleUnification(root)) { 
 // legacy logic
 }
```

Reply:
nope, we need to iterate over the `lib` and `engines` of the package paths to get the `engine-addons` of addons. Else we will only leverage whatever is in the root's `engine-addons` which might not be a complete list. If you see carefully, the `not(MOduleUnififcation)` code has the logic based on package json